diff --git b/core/lib/Drupal/Core/Form/FormPreprocess.php a/core/lib/Drupal/Core/Form/FormPreprocess.php
index 342f92f1ff..b581917a4e 100644
--- b/core/lib/Drupal/Core/Form/FormPreprocess.php
+++ a/core/lib/Drupal/Core/Form/FormPreprocess.php
@@ -351,7 +351,6 @@ public function preprocessDetails(array &$variables): void {
     $variables['attributes'] = $element['#attributes'];
     $variables['summary_attributes'] = new Attribute($element['#summary_attributes']);
     if (!empty($element['#title'])) {
-      $variables['summary_attributes']['role'] = 'button';
       if (!empty($element['#attributes']['id'])) {
         $variables['summary_attributes']['aria-controls'] = $element['#attributes']['id'];
       }
@@ -364,6 +363,7 @@ public function preprocessDetails(array &$variables): void {
       $variables['title'] = ['#markup' => $variables['title']];
     }
     $variables['description'] = (!empty($element['#description'])) ? $element['#description'] : '';
+    $variables['description_display'] = $element['#description_display'] ?? 'after';
     $variables['children'] = (isset($element['#children'])) ? $element['#children'] : '';
     $variables['value'] = (isset($element['#value'])) ? $element['#value'] : '';
     $variables['required'] = !empty($element['#required']) ? $element['#required'] : NULL;
