diff --git a/core/themes/default_admin/templates/form/form-two-columns.html.twig b/core/themes/default_admin/templates/form/form-two-columns.html.twig
index 1234567..abcdefg 100644
--- a/core/themes/default_admin/templates/form/form-two-columns.html.twig
+++ b/core/themes/default_admin/templates/form/form-two-columns.html.twig
@@ -1,6 +1,8 @@
 <div class="gin--layout-form clearfix">
   <div class="layout-region layout-region--main">
     <div class="layout-region__content">
+      <nav aria-label="{{ 'Theme switcher'|t }}">
       {% block main %}
       {% endblock %}
+      </nav>
     </div>
