diff --git b/core/modules/language/src/Hook/LanguageThemeHooks.php a/core/modules/language/src/Hook/LanguageThemeHooks.php
index 269cca04bd..79d7199033 100644
--- b/core/modules/language/src/Hook/LanguageThemeHooks.php
+++ a/core/modules/language/src/Hook/LanguageThemeHooks.php
@@ -177,6 +177,9 @@ public function preprocessLanguageContentSettingsTable(array &$variables): void
   public function preprocessBlock(&$variables): void {
     if ($variables['configuration']['provider'] == 'language') {
       $variables['attributes']['role'] = 'navigation';
+      if (empty($variables['attributes']['aria-label']) && empty($variables['attributes']['aria-labelledby'])) {
+        $variables['attributes']['aria-label'] = !empty($variables['label']) ? $variables['label'] : $this->t('Language switcher');
+      }
     }
   }
 
diff --git b/core/modules/language/tests/src/Functional/LanguageSwitchingTest.php a/core/modules/language/tests/src/Functional/LanguageSwitchingTest.php
index 825cac86af..801b148552 100644
--- b/core/modules/language/tests/src/Functional/LanguageSwitchingTest.php
+++ a/core/modules/language/tests/src/Functional/LanguageSwitchingTest.php
@@ -115,6 +115,11 @@ protected function doTestHomePageLinks(string|int|float|bool $block_label): void
     $this->DrupalGet('<front>');
     // The language switcher block should display.
     $this->assertSession()->pageTextContains($block_label);
+    $language_switcher_nav = $this->xpath('//div[@id=:id and @role="navigation" and @aria-label=:label]', [
+      ':id' => 'block-test-language-block',
+      ':label' => (string) $block_label,
+    ]);
+    $this->assertNotEmpty($language_switcher_nav, 'Language switcher block navigation landmark has accessible name.');
     // Assert that each list item and anchor element has the appropriate data-
     // attributes.
     $language_switchers = $this->xpath('//div[@id=:id]/ul/li', [':id' => 'block-test-language-block']);
