diff --git b/core/themes/default_admin/src/Helper.php a/core/themes/default_admin/src/Helper.php
index ba87af3a68..ae850242ba 100644
--- b/core/themes/default_admin/src/Helper.php
+++ a/core/themes/default_admin/src/Helper.php
@@ -60,7 +60,7 @@ public static function accentColors(): array {
       ],
       'orange' => [
         'label' => t('Orange'),
-        'hex' => '#c55228',
+        'hex' => '#bf4e25',
       ],
       'yellow' => [
         'label' => t('Yellow'),
