diff --git a/core/core.services.yml b/core/core.services.yml
index 54efa47034b..51397b165d8 100644
--- a/core/core.services.yml
+++ b/core/core.services.yml
@@ -530,6 +530,7 @@ services:
   Drupal\Core\Form\FormSubmitterInterface: '@form_submitter'
   form_error_handler:
     class: Drupal\Core\Form\FormErrorHandler
+    arguments: ['@renderer']
   Drupal\Core\Form\FormErrorHandlerInterface: '@form_error_handler'
   form_cache:
     class: Drupal\Core\Form\FormCache
diff --git a/core/lib/Drupal/Core/Form/FormErrorHandler.php b/core/lib/Drupal/Core/Form/FormErrorHandler.php
index d4124d2ea38..7d40c23869e 100644
--- a/core/lib/Drupal/Core/Form/FormErrorHandler.php
+++ b/core/lib/Drupal/Core/Form/FormErrorHandler.php
@@ -5,6 +5,7 @@ namespace Drupal\Core\Form;
 use Drupal\Component\Utility\NestedArray;
 use Drupal\Core\Messenger\MessengerTrait;
 use Drupal\Core\Render\Element;
+use Drupal\Core\Render\RendererInterface;
 
 /**
  * Handles form errors.
@@ -13,6 +14,23 @@ class FormErrorHandler implements FormErrorHandlerInterface {
 
   use MessengerTrait;
 
+  /**
+   * The renderer service.
+   *
+   * @var \Drupal\Core\Render\RendererInterface
+   */
+  protected $renderer;
+
+  /**
+   * Constructs a FormErrorHandler instance.
+   *
+   * @param \Drupal\Core\Render\RendererInterface $renderer
+   *   The renderer service.
+   */
+  public function __construct(RendererInterface $renderer) {
+    $this->renderer = $renderer;
+  }
+
   /**
    * {@inheritdoc}
    */
@@ -39,10 +57,24 @@ class FormErrorHandler implements FormErrorHandlerInterface {
    */
   protected function displayErrorMessages(array $form, FormStateInterface $form_state) {
     $errors = $form_state->getErrors();
-
-    // Loop through all form errors and set an error message.
-    foreach ($errors as $error) {
-      $this->messenger()->addMessage($error, 'error');
+    // Loop through all form errors and sets an ID & error messages.
+    foreach ($errors as $name => $error) {
+      $form_element = FormElementHelper::getElementByName($name, $form);
+      if (isset($form_element['#id'])) {
+        // Render the error messages as HTML.
+        $message = [
+          '#prefix' => '<div id="' . $form_element['#id'] . '--error-message">',
+          '#suffix' => '</div>',
+          'message' => [
+            '#markup' => $error,
+          ],
+        ];
+        $message = $this->renderer->renderInIsolation($message);
+        $this->messenger()->addMessage($message, 'error');
+      }
+      else {
+        $this->messenger()->addMessage($error, 'error');
+      }
     }
   }
 
@@ -164,6 +196,12 @@ class FormErrorHandler implements FormErrorHandlerInterface {
 
     // Store the errors for this element on the element directly.
     $elements['#errors'] = $form_state->getError($elements);
+
+    // Add aria-describedby attribute to the form element.
+    if (($elements['#errors']) !== NULL) {
+      $elements['#attributes']['aria-errormessage'] = $elements['#id'] . '--error-message';
+      $elements['#attributes']['aria-invalid'] = TRUE;
+    }
   }
 
 }
diff --git a/core/misc/cspell/dictionary.txt b/core/misc/cspell/dictionary.txt
index ff0aef453c2..d18cc04b5ba 100644
--- a/core/misc/cspell/dictionary.txt
+++ b/core/misc/cspell/dictionary.txt
@@ -163,6 +163,7 @@ endtrans
 enoki
 entityviewedit
 errmode
+errormessages
 eslintcache
 español
 eurozone
diff --git a/core/modules/inline_form_errors/tests/src/Functional/InlineErrorFormTest.php b/core/modules/inline_form_errors/tests/src/Functional/InlineErrorFormTest.php
new file mode 100644
index 00000000000..a8212cc5096
--- /dev/null
+++ b/core/modules/inline_form_errors/tests/src/Functional/InlineErrorFormTest.php
@@ -0,0 +1,112 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Tests\inline_form_errors\Functional;
+
+use Drupal\Tests\BrowserTestBase;
+
+/**
+ * Validation test for inline_form_errors.
+ *
+ * @group inline_form_errors
+ * @phpstan-ignore testClass.missingAttribute.Group, testClass.missingAttribute.RunInSeparateProcesses
+ */
+class InlineErrorFormTest extends BrowserTestBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected $defaultTheme = 'stark';
+
+  /**
+   * {@inheritdoc}
+   */
+  protected static $modules = [
+    'inline_form_errors',
+    'form_test',
+  ];
+
+  /**
+   * Tests validation for required checkbox, select, and radio elements.
+   *
+   * Submits a test form containing several types of form elements. The form
+   * is submitted twice, first without values for required fields and then
+   * with values. Each submission is checked for relevant error messages.
+   *
+   * @see \Drupal\form_test\Form\FormTestValidateRequiredForm
+   */
+  public function testRequiredCheckboxesRadio(): void {
+    $form = \Drupal::formBuilder()->getForm('\Drupal\form_test\Form\FormTestValidateRequiredForm');
+
+    // Attempt to submit the form with no required fields set.
+    $edit = [];
+    $this->drupalGet('form-test/validate-required');
+    $this->submitForm($edit, 'Submit');
+    $session = $this->assertSession();
+
+    // The only error messages that should appear are the relevant 'required'
+    // messages for each field.
+    $expected = [];
+    foreach (['textfield', 'checkboxes', 'select', 'radios'] as $key) {
+      if (isset($form[$key]['#required_error'])) {
+        $expected[] = $form[$key]['#required_error'];
+      }
+      elseif (isset($form[$key]['#form_test_required_error'])) {
+        $expected[] = $form[$key]['#form_test_required_error'];
+      }
+      else {
+        $expected[] = $form[$key]['#title'] . ' field is required.';
+      }
+      // Ensure errors are wrapped with an ID that can be referenced by aria-errormessages
+      $error_id = $form[$key]['#id'] . '--error-message';
+      $session->elementExists('css', '#' . $error_id);
+      // Make sure form element references error.
+      $type = match ($form[$key]['#type']) {
+        'textfield' => 'input',
+        'checkboxes' => 'checkbox',
+        'select' => 'select',
+        'radios' => 'radio',
+      };
+      $session->elementExists('css', $type . '[aria-errormessage=' . $error_id . ']');
+    }
+
+    // Check the page for error messages.
+    $errors = $this->xpath('//div[contains(@class, "error")]//li');
+    foreach ($errors as $error) {
+      $this->assertContains($error->getText(), array_map(fn($n) => (string) $n, $expected));
+      $expected_key = array_search($error->getText(), $expected);
+      unset($expected[$expected_key]);
+    }
+
+    // Fail if any expected messages were not found.
+    $this->assertEmpty($expected, 'Found unexpected error messages');
+
+    // Verify that input elements are still empty.
+    $session->fieldValueEquals('textfield', '');
+    $session->checkboxNotChecked('edit-checkboxes-foo');
+    $session->checkboxNotChecked('edit-checkboxes-bar');
+    $this->assertTrue($session->optionExists('edit-select', '')->isSelected());
+    $session->checkboxNotChecked('edit-radios-foo');
+    $session->checkboxNotChecked('edit-radios-bar');
+    $session->checkboxNotChecked('edit-radios-optional-foo');
+    $session->checkboxNotChecked('edit-radios-optional-bar');
+    $session->checkboxNotChecked('edit-radios-optional-default-value-false-foo');
+    $session->checkboxNotChecked('edit-radios-optional-default-value-false-bar');
+
+    // Submit again with required fields set and verify that there are no
+    // error messages.
+    $edit = [
+      'textfield' => $this->randomString(),
+      'checkboxes[foo]' => TRUE,
+      'select' => 'foo',
+      'radios' => 'bar',
+    ];
+    $this->submitForm($edit, 'Submit');
+    // Verify that no error message is displayed when all required fields are
+    // filled.
+    $this->assertSession()->elementNotExists('xpath', '//div[contains(@class, "error")]');
+    $this->assertSession()->pageTextContains("The form_test_validate_required_form form was submitted successfully.");
+  }
+
+}
diff --git a/core/modules/inline_form_errors/tests/src/Unit/FormErrorHandlerTest.php b/core/modules/inline_form_errors/tests/src/Unit/FormErrorHandlerTest.php
index ff19dd75c55..d402a7a79a8 100644
--- a/core/modules/inline_form_errors/tests/src/Unit/FormErrorHandlerTest.php
+++ b/core/modules/inline_form_errors/tests/src/Unit/FormErrorHandlerTest.php
@@ -189,8 +189,11 @@ class FormErrorHandlerTest extends UnitTestCase {
         FALSE
       );
 
-    $this->renderer->expects($this->never())
-      ->method('renderInIsolation');
+    $this->renderer->expects($this->atLeastOnce())
+      ->method('renderInIsolation')
+      ->willReturnCallback(function ($message) {
+        return $message['message']['#markup'];
+      });
 
     $this->testForm['#disable_inline_form_errors'] = TRUE;
 
diff --git a/core/modules/system/templates/fieldset.html.twig b/core/modules/system/templates/fieldset.html.twig
index f292edb23c1..8b9c32dec93 100644
--- a/core/modules/system/templates/fieldset.html.twig
+++ b/core/modules/system/templates/fieldset.html.twig
@@ -53,7 +53,7 @@
       <div{{ description.attributes.addClass('description') }}>{{ description.content }}</div>
     {% endif %}
     {% if errors %}
-      <div>
+      <div id="{{ element['#id'] }}--error-message" class="form-item--error-message">
         {{ errors }}
       </div>
     {% endif %}
diff --git a/core/modules/system/templates/form-element.html.twig b/core/modules/system/templates/form-element.html.twig
index 1c3846b6c8a..d9ae3abda9e 100644
--- a/core/modules/system/templates/form-element.html.twig
+++ b/core/modules/system/templates/form-element.html.twig
@@ -85,7 +85,7 @@
     {{ label }}
   {% endif %}
   {% if errors %}
-    <div class="form-item--error-message">
+    <div id="{{ element['#id'] }}--error-message" class="form-item--error-message">
       {{ errors }}
     </div>
   {% endif %}
diff --git a/core/modules/system/tests/src/Functional/Form/FormTest.php b/core/modules/system/tests/src/Functional/Form/FormTest.php
index 8b40dd27da2..6af400b727a 100644
--- a/core/modules/system/tests/src/Functional/Form/FormTest.php
+++ b/core/modules/system/tests/src/Functional/Form/FormTest.php
@@ -215,6 +215,7 @@ class FormTest extends BrowserTestBase {
     $edit = [];
     $this->drupalGet('form-test/validate-required');
     $this->submitForm($edit, 'Submit');
+    $session = $this->assertSession();
 
     // The only error messages that should appear are the relevant 'required'
     // messages for each field.
@@ -229,38 +230,41 @@ class FormTest extends BrowserTestBase {
       else {
         $expected[] = $form[$key]['#title'] . ' field is required.';
       }
+      // Ensure errors are wrapped with an ID that can be referenced by aria-errormessages
+      $error_id = $form[$key]['#id'] . '--error-message';
+      $session->elementExists('css', '.messages--error #' . $error_id);
+      // Make sure form element references error.
+      $type = match ($form[$key]['#type']) {
+        'textfield' => 'input',
+        'checkboxes' => 'checkbox',
+        'select' => 'select',
+        'radios' => 'radio',
+      };
+      $session->elementExists('css', $type . '[aria-errormessage=' . $error_id . ']');
     }
 
     // Check the page for error messages.
     $errors = $this->xpath('//div[contains(@class, "error")]//li');
     foreach ($errors as $error) {
+      $this->assertContains($error->getText(), array_map(fn($n) => (string) $n, $expected));
       $expected_key = array_search($error->getText(), $expected);
-      // If the error message is not one of the expected messages, fail.
-      if ($expected_key === FALSE) {
-        $this->fail("Unexpected error message: " . $error[0]);
-      }
-      // Remove the expected message from the list once it is found.
-      else {
-        unset($expected[$expected_key]);
-      }
+      unset($expected[$expected_key]);
     }
 
     // Fail if any expected messages were not found.
-    foreach ($expected as $not_found) {
-      $this->fail("Found error message: " . $not_found);
-    }
+    $this->assertEmpty($expected, 'Found unexpected error messages');
 
     // Verify that input elements are still empty.
-    $this->assertSession()->fieldValueEquals('textfield', '');
-    $this->assertSession()->checkboxNotChecked('edit-checkboxes-foo');
-    $this->assertSession()->checkboxNotChecked('edit-checkboxes-bar');
-    $this->assertTrue($this->assertSession()->optionExists('edit-select', '')->isSelected());
-    $this->assertSession()->checkboxNotChecked('edit-radios-foo');
-    $this->assertSession()->checkboxNotChecked('edit-radios-bar');
-    $this->assertSession()->checkboxNotChecked('edit-radios-optional-foo');
-    $this->assertSession()->checkboxNotChecked('edit-radios-optional-bar');
-    $this->assertSession()->checkboxNotChecked('edit-radios-optional-default-value-false-foo');
-    $this->assertSession()->checkboxNotChecked('edit-radios-optional-default-value-false-bar');
+    $session->fieldValueEquals('textfield', '');
+    $session->checkboxNotChecked('edit-checkboxes-foo');
+    $session->checkboxNotChecked('edit-checkboxes-bar');
+    $this->assertTrue($session->optionExists('edit-select', '')->isSelected());
+    $session->checkboxNotChecked('edit-radios-foo');
+    $session->checkboxNotChecked('edit-radios-bar');
+    $session->checkboxNotChecked('edit-radios-optional-foo');
+    $session->checkboxNotChecked('edit-radios-optional-bar');
+    $session->checkboxNotChecked('edit-radios-optional-default-value-false-foo');
+    $session->checkboxNotChecked('edit-radios-optional-default-value-false-bar');
 
     // Submit again with required fields set and verify that there are no
     // error messages.
diff --git a/core/tests/Drupal/Tests/Core/Form/FormErrorHandlerTest.php b/core/tests/Drupal/Tests/Core/Form/FormErrorHandlerTest.php
index eb21dd551bd..4d0ca807a3c 100644
--- a/core/tests/Drupal/Tests/Core/Form/FormErrorHandlerTest.php
+++ b/core/tests/Drupal/Tests/Core/Form/FormErrorHandlerTest.php
@@ -9,6 +9,7 @@ use Drupal\Core\Form\FormErrorHandler;
 use Drupal\Core\Form\FormState;
 use Drupal\Core\Messenger\MessengerInterface;
 use Drupal\Tests\UnitTestCase;
+use Prophecy\Argument;
 use PHPUnit\Framework\Attributes\CoversClass;
 use PHPUnit\Framework\Attributes\Group;
 
@@ -29,6 +30,13 @@ class FormErrorHandlerTest extends UnitTestCase {
    */
   protected MessengerInterface $messenger;
 
+  /**
+   * The renderer service.
+   *
+   * @var \Drupal\Core\Render\RendererInterface|\Prophecy\Prophecy\ObjectProphecy
+   */
+  protected $renderer;
+
   /**
    * {@inheritdoc}
    */
@@ -40,7 +48,7 @@ class FormErrorHandlerTest extends UnitTestCase {
     $container->set('messenger', $this->messenger);
     \Drupal::setContainer($container);
 
-    $this->formErrorHandler = new FormErrorHandler();
+    $this->formErrorHandler = new FormErrorHandler($this->renderer);
   }
 
   /**
@@ -80,6 +88,9 @@ class FormErrorHandlerTest extends UnitTestCase {
         'error',
       );
 
+    $this->renderer->renderInIsolation(Argument::any())
+      ->will(fn(array $args) => $args[0]['message']['#markup']);
+
     $form = [
       '#parents' => [],
       '#array_parents' => [],
