diff --git a/core/themes/default_admin/src/Settings.php b/core/themes/default_admin/src/Settings.php
index 505207abc4d..2b4ecf378cd 100644
--- a/core/themes/default_admin/src/Settings.php
+++ b/core/themes/default_admin/src/Settings.php
@@ -342,7 +342,7 @@ public function getSettingsForm(?AccountInterface $account = NULL): array {
     $form['high_contrast_mode'] = [
       '#type' => 'checkbox',
       '#title' => $this->t('Increase contrast') . $experimental_label,
-      '#description' => $this->t('Enables high contrast mode.'),
+      '#description' => $this->t('Increases text and interface contrast while preserving visual hierarchy.'),
       '#default_value' => $account ? $this->get('high_contrast_mode', $account) : $this->getDefault('high_contrast_mode'),
     ];
 
